website/www/site/content/en/documentation/programming-guide.md

```diff
@@ -5188,16 +5188,60 @@ restriction pairs.
 #### 12.1.1. A basic SDF {#a-basic-sdf}
 
 A basic SDF is composed of three parts: a restriction, a restriction provider, and a
-restriction tracker. The restriction is used to represent a subset of work for a given element.
-The restriction provider lets SDF authors override default implementations for splitting, sizing,
-watermark estimation, and so forth. In [Java](https://github.com/apache/beam/blob/f4c2734261396858e388ebef2eef50e7d48231a8/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java#L92)
+restriction tracker. If you want to control the watermark properly especially in a streaming
+pipeline, two more components are needed: a watermark estimator provider and a watermark estimator.
+
+The restriction is a user-defined object which is used to represent a subset of
+work for a given element. For example, we defined OffsetRange as a restriction to represent offset
+positions in [Java](https://beam.apache.org/releases/javadoc/2.25.0/org/apache/beam/sdk/io/range/OffsetRange.html) 
```

Comment:
Do we have a way to always link to the latest released java/py doc? @rosetn

Reply:
You can link to `current`. See: https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/range/OffsetRange.html

Reply:
Thanks! That's really helpful.